cli/exp_task_logs.go

```diff
@@ -71,8 +89,167 @@ func (r *RootCmd) taskLogs() *serpent.Command {
+	ctx := inv.Context()
+	workspace, err := client.Workspace(ctx, taskID)
+	if err != nil {
+		return xerrors.Errorf("fetch workspace: %w", err)
+	}
+	if workspace.LatestBuild.ID == uuid.Nil {
+		return xerrors.Errorf("workspace has no builds")
+	}
+	build := workspace.LatestBuild
+	if build.HasAITask == nil || !*build.HasAITask {
+		return xerrors.Errorf("workspace is not configured as an AI task")
+	}
+
+	if build.AITaskSidebarAppID == nil || *build.AITaskSidebarAppID == uuid.Nil {
+		return xerrors.Errorf("task is not configured with a sidebar app")
+	}
+
+	sidebarAppID := *build.AITaskSidebarAppID
+	var agentID uuid.UUID
+	var sidebarApp *codersdk.WorkspaceApp
+
+	for _, res := range build.Resources {
+		for _, agent := range res.Agents {
+			for _, app := range agent.Apps {
+				if app.ID == sidebarAppID {
+					agentID = agent.ID
+					sidebarApp = &app
+					break
+				}
+			}
+			if sidebarApp != nil {
+				break
+			}
+		}
+		if sidebarApp != nil {
+			break
+		}
+	}
+
+	if sidebarApp == nil {
+		return xerrors.Errorf("task sidebar app not found in latest build")
+	}
+	switch sidebarApp.Health {
+	case codersdk.WorkspaceAppHealthInitializing:
+		return xerrors.Errorf("task sidebar app is initializing, try again shortly")
+	case codersdk.WorkspaceAppHealthUnhealthy:
+		return xerrors.Errorf("task sidebar app is unhealthy")
+	}
+	if sidebarApp.URL == "" {
+		return xerrors.Errorf("task sidebar app URL is not configured")
+	}
+	parsedURL, err := url.Parse(sidebarApp.URL)
+	if err != nil {
+		return xerrors.Errorf("parse task app URL: %w", err)
+	}
+	if parsedURL.Scheme != "http" {
+		return xerrors.Errorf("only http scheme is supported for direct agent-dial")
+	}
+
+	cliui.Infof(inv.Stderr, "Connecting to task workspace agent...")
+	dialCtx, dialCancel := context.WithTimeout(ctx, 30*time.Second)
+	defer dialCancel()
+
+	agentConn, err := workspacesdk.New(client).DialAgent(dialCtx, agentID, &workspacesdk.DialAgentOptions{
+		Logger: inv.Logger,
+	})
+	if err != nil {
+		return xerrors.Errorf("dial workspace agent: %w", err)
+	}
+	defer agentConn.Close()
+
+	// Wait for the connection to be reachable
+	agentConn.AwaitReachable(ctx)
+
+	// Create HTTP client that dials through the agent
+	httpClient := &http.Client{
+		Transport: &http.Transport{
+			DialContext: func(ctx context.Context, network, addr string) (net.Conn, error) {
+				return agentConn.DialContext(ctx, network, addr)
+			},
+		},
+	}
+
+	agentAPIClient, err := aiagentapi.NewClient(parsedURL.String(), aiagentapi.WithHTTPClient(httpClient))
+	if err != nil {
+		return xerrors.Errorf("create agent API client: %w", err)
+	}
+
+	cliui.Infof(inv.Stderr, "Subscribing to log events...")
+	eventsCh, errCh, err := agentAPIClient.SubscribeEvents(ctx)
+	if err != nil {
+		return xerrors.Errorf("subscribe to events: %w", err)
+	}
+
+	messagesResp, err := agentAPIClient.GetMessages(ctx)
+	if err != nil {
+		return xerrors.Errorf("get existing messages: %w", err)
+	}
```

Comment:
I could be misremembering but I thought agent API's `/events` replayed existing messages already?  I wonder if it needs an `after` flag, come to think of it.

Reply:
Yeah I had a long comment prepped here and got distracted but you end up having to handle a load of message update events, meaning clearing and re-drawing the terminal every so often. I haven't gotten this working quite right yet. 😿